### PR TITLE
SNOW-2235948: Adding MFA tests to scope

### PR DESCRIPTION
**Pull request by sfc-gh-pcyrek**
### Description

SNOW-2235948: Adding MFA tests to scope

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: auth_with_mfa_test.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	totpKeys := getTOPTcodes(t)
+	log.Printf("Got %d TOTP codes to try", len(totpKeys))
+
+	err := verifyConnectionToSnowflakeUsingTotpCodes(t, cfg, totpKeys)
+	if err != nil {
+		t.Fatalf("Failed to connect with any of the %d TOTP codes: %v", len(totpKeys), err)
+		return
+	}
+
+	log.Printf("Testing MFA token caching with second connection...")
```
Comment (sfc-gh-akolodziejczyk) [inline auth_with_mfa_test.go:23]: I dont think we need to print those logs, assertion and errors should be enogh

**Review by sfc-gh-akolodziejczyk — APPROVED** [review]

File: auth_with_mfa_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	// Enable MFA token caching
+	cfg.ClientRequestMfaToken = ConfigBoolTrue
+
+	//Provide your own TOTP code/codes here, to test manually
+	//totpKeys := []string{"222222", "333333", "444444"}
+
+	totpKeys := getTOPTcodes(t)
+
+	err := verifyConnectionToSnowflakeUsingTotpCodes(t, cfg, totpKeys)
+	if err != nil {
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go]: `assertNilF`?

File: auth_with_mfa_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		t.Fatalf("Failed to connect with any of the %d TOTP codes: %v", len(totpKeys), err)
+		return
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go]: Don't need to return if fatal is used - fatal breaks execution.

File: auth_with_mfa_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	log.Printf("Testing MFA token caching with second connection...")
+
+	// Clear the passcode to force use of cached MFA token
+	cfg.Passcode = ""
+
+	// Attempt to connect using cached MFA token
+	cacheErr := verifyConnectionToSnowflakeAuthTests(t, cfg)
+	if cacheErr != nil {
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go]: Same.

File: auth_with_mfa_test.go (hunk 1 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go:23]: This one has value, it instantly tells you if the issue is with caching or connection in general.

File: auth_with_mfa_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		t.Fatalf("Failed to connect with cached MFA token: %v", cacheErr)
+		return
+	}
+}
+
+func setupMfaTest(t *testing.T) *Config {
+	skipAuthTests(t, "Skipping MFA tests")
+	cfg, err := getAuthTestsConfig(t, AuthTypeUsernamePasswordMFA)
+	assertEqualE(t, err, nil, fmt.Sprintf("failed to get config: %v", err))
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go]: `assertNilF`? Also, can use `E`, in case of `E` a test continues, so it will panic on nil config on the next line.

File: auth_with_mfa_test.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	assertNilF(t, err, fmt.Sprintf("failed to setup config: %v", err))
+
+	return cfg
+}
+
+func getTOPTcodes(t *testing.T) []string {
+	if isTestRunningInDockerContainer() {
+		const provideTotpPath = "/externalbrowser/totpGenerator.js"
+		output, err := exec.Command("node", provideTotpPath).CombinedOutput()
+		assertNilE(t, err, fmt.Sprintf("failed to execute command: %v", err))
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go]: Same.

File: auth_with_mfa_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+		totpCodes := strings.Fields(string(output))
+		return totpCodes
+	}
+	return []string{}
+}
+
+func verifyConnectionToSnowflakeUsingTotpCodes(t *testing.T, cfg *Config, totpKeys []string) error {
+	if len(totpKeys) == 0 {
+		return fmt.Errorf("no TOTP codes provided")
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go]: `errors.New`? Or even better - `t.Fatal`?

File: auth_with_mfa_test.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+			log.Printf("MFA error detected, trying next code...")
+			continue
+		} else {
+			log.Printf("Non-MFA error detected: %v", errorMsg)
+			break
+		}
+	}
+
+	if lastError != nil {
+		return fmt.Errorf("failed to connect with any TOTP code: %v", lastError)
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go]: Same here and below.

File: auth_with_mfa_test.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+	for i, totpKey := range totpKeys {
+		cfg.Passcode = totpKey
+
+		err := verifyConnectionToSnowflakeAuthTests(t, cfg)
+		if err == nil {
+			return nil
+		}
+
+		lastError = err
```
Comment (sfc-gh-pfus) [inline auth_with_mfa_test.go:73]: Optional - maybe it would be better to use error wrapping here?

File: ci/test_authentication.sh [hunk 10]
```diff
@@ -27,5 +27,5 @@ docker run \
   -v $(cd $THIS_DIR/.. && pwd):/mnt/host \
   -v $WORKSPACE:/mnt/workspace \
   --rm \
-  nexus.int.snowflakecomputing.com:8086/docker/snowdrivers-test-external-browser-golang:6 \
+  nexus.int.snowflakecomputing.com:8086/docker/snowdrivers-test-external-browser-golang:8 \
```
Comment (sfc-gh-pfus) [inline ci/test_authentication.sh:30]: Wait... I updated the image to version 8 as well. Have you overriden my image in Nexus? :/

Comment (sfc-gh-pcyrek) [reply ci/test_authentication.sh:30]: nope, my changes on :7 are already on yours

File: auth_with_mfa_test.go (hunk 2 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go]: ok

File: auth_with_mfa_test.go (hunk 3 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go]: ok

File: auth_with_mfa_test.go (hunk 4 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go]: ok

File: auth_with_mfa_test.go (hunk 7 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go]: ok

File: auth_with_mfa_test.go (hunk 6 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go]: ok

File: auth_with_mfa_test.go (hunk 8 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go]: ok

File: auth_with_mfa_test.go (hunk 5 above)
Comment (sfc-gh-pcyrek) [reply auth_with_mfa_test.go]: ok

**Review by sfc-gh-pfus — APPROVED** [review]
